presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -68,74 +74,98 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
         }, value);
     }
 
-    public AggregationNode map(AggregationNode node, PlanNode source)
+    public <T extends PlanNode> T map(PlanNode planNode, PlanNodeIdAllocator idAllocator, RecursionStrategy recursionStrategy)
```

Comment:
> I don't think SymbolMapper needs to expose RecursionStrategy to its user. Recursive uses can use generic function like this patch has, and the previous interface already worked for non-recursive uses.

No problem. I can hide it Also I will restore the previous API.

> Could you explain how the SymbolMapper builder and class is a better idea than having a static function for each Node type that takes a "fromTo" map as an explicit argument? The only thing I can come up with is that IntelliJ can prompt the user of the builder with the argument names "from" and "to", but maybe I'm missing something.

`SymbolMapper` has public constructor so you do not need to use `Builder` if you do not need. To me it is clear for example in case when you already have mapping (`fromTo` map) using a `Builder` just to create a `SymbolMapper` would be at least cumbersome. However, in case where I need to do some traversal in order to collect mapping, then it is a bit easier and cleaner to collect them in such builder than collecting them in `ImmutableMap.Builder` and then pass it to `SymbolMapper`

I would prefer to avoid to put overloaded methods, one per each `PlanNode` type (times two (or three): recursive, non recursive (with singleton and list)). Because  there is a visitor used underneath and also because it is a lot of boilerplate code. The only reason I see for now is to forbid to call a `map` with a node which is not yet supported. Is this your point?

Also please take a look why this refactoring is needed.
https://github.com/prestodb/presto/pull/8435
https://github.com/prestodb/presto/pull/8435/commits/d5720446eb4ecfbda2872ae8b7886168472e509b#diff-ab20e8b431228e8711be04b33c0acbbfR168